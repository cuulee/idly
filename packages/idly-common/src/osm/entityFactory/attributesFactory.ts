import { Attributes } from '../structures';

export function attributesFactory({
  visible,
  version,
  timestamp,
  changeset,
  uid,
  user,
}: {
  visible?: boolean;
  version?: string;
  timestamp?: string;
  changeset?: string;
  uid?: string;
  user?: string;
} = {}): Attributes {
  return {
    changeset,
    timestamp,
    uid,
    user,
    version,
    visible,
  };
}
