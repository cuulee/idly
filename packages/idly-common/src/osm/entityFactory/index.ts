export { attributesFactory } from './attributesFactory';
export { nodeFactory } from './nodeFactory';
export { relationFactory } from './relationFactory';
export { tagsFactory } from './tagsFactory';
export { wayFactory } from './wayFactory';
