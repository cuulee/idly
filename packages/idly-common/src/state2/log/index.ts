import { weakCache } from '../../misc/weakCache';
import { setCreate, setFind, setIntersect, setSome } from '../helper';
/**
 * The log is used as [...modifiedEntries, virginOsmEntries]
 * the pointer to virginOsmEntries resolves to EntityTable
 * we carefully mutate virginOsmEntries keeping the pointer same.
 *
 * modified entities are id#version where version [0, infinity) (0 is inclusive)
 */

export type Index = string;

export type Log = ReadonlyArray<Entry>;

export type Entry = ReadonlySet<string>;

export const validId = (str: string) => str.indexOf('#') === 0;

export const validIndex = (str: string) => !validId(str);

export const logGetModifiedIds = weakCache((log: Log) =>
  log.reduce((prev: Set<string>, cur) => {
    cur.forEach(index => prev.add(getId(index)));
    return prev;
  }, setCreate())
);

export const logGetLatestIndexes = weakCache(
  (log: Log) =>
    new Set(
      log
        .reduce((prev, cur) => {
          cur.forEach(
            index => !prev.has(getId(index)) && prev.set(getId(index), index)
          );
          return prev;
        }, new Map<string, string>())
        .values()
    )
);

export const logRecreate = (...entries: Entry[]) =>
  entries
    .slice(0)
    .reverse()
    .reduce((l, e) => {
      const newLog = addEntryToLog(e)(l);
      if (l === newLog) {
        throw new Error('malformed entries');
      }
      return newLog;
    }, logCreate());

export const logCreate = (): Log => Object.freeze([]);

export const doesLogContain = (entry: Entry) => (log: Log) =>
  log.some(e => setIntersect(entry, e).size > 0);

export const addEntryToLog = (entry: Entry) => (log: Log) =>
  setSome(index => doesLogHaveNewerOrCurrentVersion(index)(log), entry)
    ? log
    : Object.freeze([entry, ...log]);

export const doesLogHaveNewerOrCurrentVersion = (index: string) => (log: Log) =>
  getLatestVersion(index)(log) >= getVersion(index);

export const getVersion = (index: Index) => parseInt(parseIndex(index)[1], 10);

export const getLatestVersion = (id: string) => (log: Log) => {
  const finder = findInEntry(getId(id));
  let index;
  for (const entry of log) {
    index = finder(entry);
    if (index) {
      return getVersion(index);
    }
  }
  return -1;
};

export const genIndex = (id: string) => (log: Log) =>
  `${getId(id)}#${getLatestVersion(id)(log) + 1}`;

export const findInEntry = (indexOrId: string) => (entry: Entry) =>
  setFind(i => getId(i) === indexOrId, entry);

export const getId = (index: Index) => parseIndex(index)[0];

const parseIndex = (index: Index) => index.split('#');

export const incrVersion = (index: Index) =>
  `${getId(index)}#${getVersion(index) + 1}`;
