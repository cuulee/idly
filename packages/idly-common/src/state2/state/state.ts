import { Table } from '../table';
import { tableAdd, tableCreate, tableHas } from '../table/regular';

import {
  QuadkeysTable,
  quadkeysTableAdd,
  quadkeysTableCreate,
  quadkeysTableCreateFrom,
  quadkeysTableFindRelated,
  quadkeysTableFlatten,
} from './quadkeysTable';

export class State<T> {
  public static create<T>(
    elementTable: Table<T> = new Map(),
    quadkeysTable = quadkeysTableCreate()
  ) {
    return new State<T>(elementTable, quadkeysTable);
  }

  // tslint:disable-next-line:variable-name
  private _elementTable: Table<T>;
  // tslint:disable-next-line:variable-name
  private _quadkeysTable: QuadkeysTable;

  private constructor(elementTable: Table<T>, quadkeysTable: QuadkeysTable) {
    this._elementTable = elementTable;
    this._quadkeysTable = quadkeysTable;
  }
  public getElementTable() {
    return this._elementTable;
  }
  public getQuadkeysTable() {
    return this._quadkeysTable;
  }

  public getElement(id: string) {
    return this._elementTable.get(id);
  }

  public getIdByQuadkey(quadkey: string) {
    return this._quadkeysTable.get(quadkey);
  }

  public has(quadkey: string) {
    return this._quadkeysTable.has(quadkey);
  }

  // a single element can be on multiple quadkeys, but
  // changing an existing element is disallowed
  public add(getId: (t: T) => string, elements: T[], quadkey: string) {
    elements.forEach(
      e =>
        tableHas(this._elementTable, getId(e)) ||
        tableAdd(this._elementTable, getId(e), e)
    );
    quadkeysTableAdd(this._quadkeysTable, elements.map(getId), quadkey);
  }

  public getQuadkey(quadkey: string) {
    return this._quadkeysTable.get(quadkey);
  }
  public getVisible(quadkeys: string[]) {
    return quadkeysTableFindRelated(this._quadkeysTable, quadkeys);
  }

  public shred(quadkey: string) {
    const newElementTable = tableCreate<T>();

    const newQuadkeysTable = quadkeysTableCreateFrom(
      this._quadkeysTable,
      quadkey
    );

    quadkeysTableFlatten(newQuadkeysTable).forEach(k =>
      newElementTable.set(k, this._elementTable.get(k) as T)
    );

    return State.create(newElementTable, newQuadkeysTable);
  }
}
