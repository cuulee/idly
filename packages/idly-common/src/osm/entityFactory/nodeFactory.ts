import { lngLatFactory } from '../../geo/lngLatFactory';
import { deepFreeze } from '../../misc/deepFreeze';

import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function nodeFactory(
  {
    id,
    tags = tagsFactory(),
    loc = lngLatFactory([0, 0]),
    attributes = attributesFactory(),
  }: {
    id: EntityId;
    tags?: Tags;
    loc?: LngLat;
    attributes?: Attributes;
  },
  freeze = true
): Node {
  return deepFreeze<Node>(
    {
      attributes: attributesFactory(attributes),
      id,
      loc,
      tags: tagsFactory(tags),
      type: EntityType.NODE,
    },
    freeze
  );
}
