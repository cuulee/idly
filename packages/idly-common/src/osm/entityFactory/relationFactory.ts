import { deepFreeze } from '../../misc/deepFreeze';
import {
  Attributes,
  EntityId,
  EntityType,
  Relation,
  RelationMember,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function relationFactory(
  {
    id,
    tags = tagsFactory(),
    attributes = attributesFactory(),
    members = [],
  }: {
    id: EntityId;
    tags?: Tags;
    attributes?: Attributes;
    members?: RelationMember[] | ReadonlyArray<RelationMember>;
  },
  freeze = true
): Relation {
  return deepFreeze<Relation>(
    {
      attributes: attributesFactory(attributes),
      id,
      members,
      tags: tagsFactory(tags),
      type: EntityType.RELATION,
    },
    freeze
  );
}
