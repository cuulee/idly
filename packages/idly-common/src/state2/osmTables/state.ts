import { Log, logGetLatestIndexes, logGetModifiedIds } from '../log';
import { Table } from '../table';
import { tableAdd, tableFilter } from '../table/regular';

import {
  QuadkeysTable,
  quadkeysTableAdd,
  quadkeysTableCreate,
  quadkeysTableCreateFrom,
  quadkeysTableFindVirginIds,
  quadkeysTableFlatten,
} from './quadkeysTable';

export class State<T> {
  public static create<T>(
    elementTable: Table<T> = new Map(),
    quadkeysTable = quadkeysTableCreate()
  ) {
    return new State<T>(elementTable, quadkeysTable);
  }

  // tslint:disable-next-line:variable-name
  private _elementTable: Table<T>;
  // tslint:disable-next-line:variable-name
  private _quadkeysTable: QuadkeysTable;

  private constructor(elementTable: Table<T>, quadkeysTable: QuadkeysTable) {
    this._elementTable = elementTable;
    this._quadkeysTable = quadkeysTable;
  }
  public getElementTable() {
    return this._elementTable;
  }
  public getQuadkeysTable() {
    return this._quadkeysTable;
  }

  public getElement(id: string) {
    return this._elementTable.get(id);
  }

  public getIdByQuadkey(quadkey: string) {
    return this._quadkeysTable.get(quadkey);
  }

  public add(getId: (t: T) => string, elements: T[], quadkey: string) {
    elements.forEach(e => tableAdd(this._elementTable, getId(e), e));
    quadkeysTableAdd(this._quadkeysTable, elements.map(getId), quadkey);
  }

  public getVisible(quadkeys: string[], log: Log) {
    const insideQuadkeys = quadkeysTableFindVirginIds(
      this._quadkeysTable,
      quadkeys
    );

    const toRemoveIds = logGetModifiedIds(log);

    for (const id of toRemoveIds) {
      insideQuadkeys.delete(id);
    }

    for (const id of logGetLatestIndexes(log)) {
      insideQuadkeys.add(id);
    }

    return insideQuadkeys;
  }

  public shred(quadkey: string) {
    // To test we dont remove '' key
    const shreddedQuadkeyTable = quadkeysTableCreateFrom(
      this._quadkeysTable,
      quadkey
    );

    // gets all indexes
    const shreddedIndexes = quadkeysTableFlatten(shreddedQuadkeyTable);

    const shreddedElementTable = tableFilter(
      (_, id) => shreddedIndexes.has(id),
      this._elementTable
    );

    this._elementTable = shreddedElementTable;
    this._quadkeysTable = shreddedQuadkeyTable;
  }
}
