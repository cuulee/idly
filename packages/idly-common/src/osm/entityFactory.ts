import { nodeFactory } from '../osm/nodeFactory';
import { relationFactory } from '../osm/relationFactory';
import { EntityType, Tags } from '../osm/structures';
import { Entity, Node, Relation, Way } from './structures';
import { tagsFactory } from './tagsFactory';
import { wayFactory } from './wayFactory';

export function entityFactory(entity: any): Entity {
  if (!entity.id) {
    throw new Error('not a valid entity');
  }

  if (typeof entity.tags === 'string') {
    entity.tags = tagsFactory(JSON.parse(entity.tags));
  } else if (Array.isArray(entity.tags)) {
    entity.tags = tagsFactory(entity.tags);
  } else if (!entity.hasOwnProperty('tags')) {
    entity.tags = tagsFactory([]);
  }

  if (!(entity.tags instanceof Tags)) {
    throw new Error('not proper tags');
  }

  switch (entity.type) {
    case EntityType.NODE:
      return nodeFactory(entity);

    case EntityType.WAY:
      return wayFactory(entity);

    case EntityType.RELATION:
      return relationFactory(entity);

    default:
      throw new Error('not a valid entity');
  }
}
