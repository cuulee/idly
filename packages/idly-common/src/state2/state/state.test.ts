import { nodeFactory, wayFactory } from '../../osm/entityFactory/index';
import { setCreate } from '../helper';
import { addEntryToLog, logCreate } from '../log/index';

import {
  OsmElement,
  osmStateAddModifieds,
  osmStateAddVirgins,
} from '../osmState/osmTable';
import { State } from './state';

const mapFromObj = <T>(o: any = {}): Map<string, T> =>
  Object.keys(o).reduce((prev, k) => {
    prev.set(k, o[k]);
    return prev;
  }, new Map());

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});
const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

describe('constructing state and basic tests', () => {
  it('should create', () => {
    expect(State.create()).toEqual({
      _elementTable: mapFromObj(),
      _quadkeysTable: mapFromObj(),
    });
  });
  it('should create with given values', () => {
    const state = State.create<OsmElement>();
    osmStateAddVirgins(state, [n1, n2, w1], '123');
    expect(state).toMatchSnapshot();
  });
});

describe('getVisible', () => {
  it('should get ids', () => {
    const baseSetup = () => {
      const s = State.create<OsmElement>();
      osmStateAddVirgins(s, [n1, n2, w1], '123');
      return s;
    };

    const log1 = logCreate();

    const state = baseSetup();
    osmStateAddVirgins(state, [n3], '121');
    osmStateAddVirgins(state, [n4], '130');

    const n1Hash0 = nodeFactory({
      id: 'n1#0',
    });
    const n2Hash0 = nodeFactory({
      id: 'n2#0',
    });
    const w1Hash0 = wayFactory({
      id: 'w1#0',
      nodes: [n1Hash0.id, n2Hash0.id],
    });

    const log2 = addEntryToLog(
      setCreate([n1Hash0, n2Hash0, w1Hash0].map(r => r.id))
    )(log1);

    osmStateAddModifieds(state, log2, [n1Hash0, n2Hash0, w1Hash0]);

    expect(state.getVisible(['123'], log2)).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0].map(e => e.id))
    );

    expect(state.getVisible(['121'], log2)).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0, n3].map(e => e.id))
    );

    expect(state.getVisible(['130'], log2)).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0, n4].map(e => e.id))
    );

    expect(state.getVisible(['1'], log2)).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0, n3, n4].map(e => e.id))
    );
  });
});
