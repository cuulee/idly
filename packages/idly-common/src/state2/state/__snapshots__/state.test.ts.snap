// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructing state and basic tests should create with given values 1`] = `
State {
  "_elementTable": Map {
    "n1" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "n2" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "w1" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
  },
  "_quadkeysTable": Map {
    "123" => Set {
      "n1",
      "n2",
      "w1",
    },
  },
}
`;

exports[`shred shouldnt change existing state 1`] = `
State {
  "_elementTable": Map {
    "w3" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w3",
      "nodes": Array [
        "n3",
        "n4",
        "n5",
      ],
      "tags": Object {},
      "type": "way",
    },
    "r2" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "r2",
      "members": Array [
        Object {
          "id": "w1",
          "ref": "w1",
        },
        Object {
          "id": "w3",
          "ref": "w3",
        },
        Object {
          "id": "n6",
          "ref": "n6",
        },
      ],
      "tags": Object {},
      "type": "relation",
    },
  },
  "_quadkeysTable": Map {
    "100231121" => Set {
      "w3",
      "r2",
    },
  },
}
`;
